--- Day 3: Gear Ratios ---

You and the Elf eventually reach a gondola lift station; he says
the gondola lift will take you up to the water source, but this
is as far as he can bring you. You go inside.

It doesn't take long to find the gondolas, but there seems to be
a problem: they're not moving.

"Aaah!"

You turn around to see a slightly-greasy Elf with a wrench and a
look of surprise. "Sorry, I wasn't expecting anyone! The gondola
lift isn't working right now; it'll still be a while before I can
fix it." You offer to help.

The engineer explains that an engine part seems to be missing from
the engine, but nobody can figure out which one. If you can add up
all the part numbers in the engine schematic, it should be easy to
work out which part is missing.

The engine schematic (your puzzle input) consists of a visual
representation of the engine. There are lots of numbers and
symbols you don't really understand, but apparently any number
adjacent to a symbol, even diagonally, is a "part number" and
should be included in your sum. (Periods (.) do not count as a
symbol.)

Here is an example engine schematic:

```
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
```

In this schematic, two numbers are not part numbers because they
are not adjacent to a symbol: 114 (top right) and 58 (middle
right). Every other number is adjacent to a symbol and so is a
part number; their sum is 4361.

Of course, the actual engine schematic is much larger. What is
the sum of all of the part numbers in the engine schematic?

To begin, get your puzzle input.

Answer: 

In [303]:
import re

data_example = [
    "467..114..",
    "...*......",
    "..35..633.",
    "......#...",
    "617*......",
    ".....+.58.",
    "..592.....",
    "......755.",
    "...$.*....",
    ".664.598.."
]


data_example_2 = [
"12.......*..",
"+.........34",
".......-12..",
"..78........",
"..*....60...",
"78..........",
".......23...",
"....90*12...",
"............",
"2.2......12.",
".*.........*",
"1.1.......56"]

true_data = [line.strip() for line in open("../data/aoc_2023_data_03.txt", 'r')]

data = true_data

# First, we'll determine the line numbers and positions of the symbols.
# A symbol is not a digit, and not a period.



# fint the list of symbols.
# The symbols are not digits, and not periods.

symbols = "[^0-9.]"

symbol_positions = []
for line_number, line in enumerate(data):
    for match in re.finditer(symbols, line):
        symbol_positions.append((line_number, match.start()))



In [304]:
# Now, let's read the numbers, their lines and starting positions

numbers = []
for line_number, line in enumerate(data):
    for match in re.finditer("[0-9]+", line):
        numbers.append((line_number, match.start(), match.start() + len(match.group())-1, int(match.group())))



In [305]:
# We get the cross product between symbol positions and found numbers
all_combinations = [(a,b) for a in symbol_positions for b in numbers ]

# We only consider those numbers that are within one position of the symbol position.
# a[0], a[1] is the line and columns of the symbol, 
# b[0], b[1], b[2] is the line, start and end position of the number.

feasible_part_numbers = [ 
    (a, b) for a,b in all_combinations 
        if (b[0] >= a[0]-1 and b[0] <= a[0]+1) 
                and (b[1] <= a[1]+1 and b[1] >= a[1]-1 
                        or b[2] <= a[1]+1 and b[2] >= a[1]-1) ]



In [306]:
# Let's sum the feasible part number, but only distinct values
sum([b[3] for a,b in feasible_part_numbers])

549908

## --- Part Two ---

The engineer finds the missing part and installs it in the engine!
As the engine springs to life, you jump in the closest gondola,
finally ready to ascend to the water source.

You don't seem to be going very fast, though. Maybe something is
still wrong? Fortunately, the gondola has a phone labeled "help",
so you pick it up and the engineer answers.

Before you can explain the situation, she suggests that you look
out the window. There stands the engineer, holding a phone in one
hand and waving with the other. You're going so slowly that you
haven't even left the station. You exit the gondola.

The missing part wasn't the only issue - one of the gears in the
engine is wrong. A gear is any * symbol that is adjacent to
exactly two part numbers. Its gear ratio is the result of
multiplying those two numbers together.

This time, you need to find the gear ratio of every gear and add
them all up so that the engineer can figure out which gear needs
to be replaced.

Consider the same engine schematic again:

```
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
```

In this schematic, there are two gears. The first is in the top left;
it has part numbers 467 and 35, so its gear ratio is 16345. The second
gear is in the lower right; its gear ratio is 451490. (The * adjacent
to 617 is not a gear because it is only adjacent to one part number.)

Adding up all of the gear ratios produces 467835.

What is the sum of all of the gear ratios in your engine schematic?


In [307]:
# We actually need to see which * symbols have more than one part number associated


# the gear symbol is a "*" character, so we need to escape it
gear_symbol = "\*"

gear_positions = []
for line_number, line in enumerate(data):
    for match in re.finditer(gear_symbol, line):
        gear_positions.append((line_number, match.start()))



In [308]:
# We get the cross product between gear positions and found numbers
all_combinations2 = [(a,b) for a in gear_positions for b in numbers ]

# We only consider those numbers that are within one position of the symbol position.
# a[0], a[1] is the line and columns of the symbol, 
# b[0], b[1], b[2] is the line, start and end position of the number.

feasible_part_numbers2 = [ 
    (a, b) for a,b in all_combinations2 
        if (b[0] >= a[0]-1 and b[0] <= a[0]+1) 
                and (b[1] <= a[1]+1 and b[1] >= a[1]-1 
                        or b[2] <= a[1]+1 and b[2] >= a[1]-1) ]


In [309]:
# We need to count the number of instances in feasible_part_numbers2, that has mor than one instance
# regardless of part number

gear_values = {}
for fp in feasible_part_numbers2:
    if fp[0] in gear_values:
        gear_values[fp[0]].append(fp[1][3])
    else:
        gear_values[fp[0]] = [fp[1][3]]


In [310]:
sum([ gear_values[g][0] * gear_values[g][1] for g in gear_values.keys() if len(gear_values[g]) == 2])

81166799

**Answer: 81166799**

That's the right answer! You are one gold star closer to restoring snow operations.

You have completed Day 3!